pkg/bpf/perf_test.go

```diff
@@ -0,0 +1,63 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bpf
+
+import (
+	"time"
+	"unsafe"
+
+	. "gopkg.in/check.v1"
+)
+
+func (s *BPFTestSuite) TestPerfRead(c *C) {
+	var buf [256]byte
+	var state [128]byte
+
+	var checkSet int
+
+	var x ReceiveFunc = func(msg *PerfEventSample, cpu int) {}
+	var y LostFunc = func(msg *PerfEventLost, cpu int) {}
+	var z ErrorFunc = func(msg *PerfEvent) { checkSet = 666 }
+
+	// nonsensical perf event which only aim is to loop PerfEvent.Read
+	event := PerfEvent{
+		cpu:      1,
+		Fd:       1,
+		pagesize: 8,
+		npages:   5,
+		lost:     0,
+		unknown:  0,
+		buf:      buf,
+		state:    unsafe.Pointer(&state[0]),
+	}
+
+	event.data = getTestHeader()
+
+	readDone := make(chan struct{})
+
+	go func() {
+		event.Read(x, y, z)
```

Comment:
Minor nit and one bug (I think, please confirm): for the nit, could we move the `getTestHeader()` and C-based `get_test_header()` also into perf_test.go? The latter C function doesn't perform NULL check and neither free()s the mem (though just minor here). However, I believe there's an out-of-bounds bug here: `event.data` is of size `struct perf_event_mmap_page` which you populate earlier. Setting this as `event.data` (which is in real life 1 page with head/tail followed by one or more pages of data content that is mmaped) and calling `event.Read()` will access this memory out of bounds e.g. when you walk the buffer in `perf_event_read()`. Depending on the malloc'ed content after the `perf_event_mmap_page` we'd try to walk garbage and could potentially segfault in the test.

Reply:
(Aside from the test, the rest looks good to me.)

Reply:
@nebril Probably easiest to just remove the test case for now and follow up on it later on.